experimental/slo/slo_middleware.go

```diff
@@ -0,0 +1,95 @@
+package slo
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var duration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: "grafana",
+	Name:      "plugin_external_requests_duration_seconds",
+	Help:      "Duration of requests to external services",
+}, []string{"datasource_name", "datasource_type", "error_source"})
+
+const DataSourceSLOMiddlewareName = "slo"
+
+// Middleware captures duration of requests to external services and the source of errors
+func Middleware(plugin string) httpclient.Middleware {
+	return httpclient.NamedMiddlewareFunc(DataSourceSLOMiddlewareName, func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
+		return RoundTripper(plugin, opts, next)
+	})
+}
+
+// RoundTripper captures duration of requests to external services and the source of errors
+func RoundTripper(plugin string, opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
+	return httpclient.RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
+		start := time.Now()
+		var errorSource = "none"
+
+		defer func() {
+			if opts.Labels == nil {
+				return
+			}
+
+			datasourceName, exists := opts.Labels["datasource_name"]
+			if !exists {
+				return
+			}
+
+			datasourceLabelName, err := SanitizeLabelName(datasourceName)
```

Comment:
Nit: I wonder whether this may have an impact on performance given it's run for every request and builds a new string by iterating over all chars each time? Probably not but maybe worth keeping an eye out for this but I'd say it's fine for now especially since it's in `experimental` and opt-in :slightly_smiling_face:

Reply:
Well spotted, the label extraction can actually happen before/outside `return httpclient.RoundTripperFunc`. We're actually doing just that here for reference https://github.com/grafana/grafana/blob/5c27f223aff8d3d1db1eb85d178e5ad86dcc534b/pkg/infra/httpclient/httpclientprovider/datasource_metrics_middleware.go#L59-L92